lib/spack/spack/directory_layout.py

```diff
@@ -74,6 +75,13 @@ def path_for_spec(self, spec):
         """Return absolute path from the root to a directory for the spec."""
         _check_concrete(spec)
 
+        if spec.external:
+            return spec.external_path
+        if self.check_upstream and spec.package.installed_upstream:
+            raise SpackError(
```

Comment:
Why not just return the upstream path here?

Reply:
My philosophy is that this function should not be called for upstream-installed packages and doing so implies there is an error in the Spack core.